src/test/java/com/flagsmith/FlagsmithCachedClientTest.java

```diff
@@ -0,0 +1,356 @@
+
+    // Act: get flags from cache instead of reading new values from API
+    final boolean newFlagValue = !user2feature1initValue;
+    switchFlagForUser(featureId, userId2, newFlagValue, environment.apiKey);
+    assignTraitToUserIdentity("mr-user-2", "trait2", "new-trait-val", environment.apiKey);
+    final FlagsAndTraits flagsAndTraits2Again = environment.client.getUserFlagsAndTraits(user2);
+    assertEquals(2, clientCache.estimatedSize());
+    assertEquals(flagsAndTraits2, flagsAndTraits2Again);
+
+    assertNotNull(flagsAndTraits2);
+    assertThat(flagsAndTraits2.getFlags())
+        .hasSize(2)
+        .containsExactlyInAnyOrder(
+            flag("Flag to be enabled for the user", null, true),
+            flag("Other Flag", null, false)
+        );
+    assertThat(flagsAndTraits2.getTraits())
+        .hasSize(1)
+        .contains(trait(null, "foo1", "xxx"));
+
+    // Act: clean cache should fetch new traits
+    clientCache.invalidate("mr-user-2");
+    assertEquals(1, clientCache.estimatedSize());
+    flagsAndTraits2 = environment.client.getUserFlagsAndTraits(user2);
+    assertEquals(2, clientCache.estimatedSize());
+    assertNotEquals(flagsAndTraits2, flagsAndTraits2Again);
+
+    assertNotNull(flagsAndTraits2);
+    assertThat(flagsAndTraits2.getFlags())
+        .hasSize(2)
+        .containsExactlyInAnyOrder(
+            flag("Flag to be enabled for the user", null, newFlagValue),
+            flag("Other Flag", null, false)
+        );
+    assertThat(flagsAndTraits2.getTraits())
+        .hasSize(2)
+        .containsExactlyInAnyOrder(
+            trait(null, "foo1", "xxx"),
+            trait(null, "trait2", "new-trait-val")
+        );
+  }
+
+  @Test(groups = "integration")
+  public void testClient_When_Get_User_Trait_Update_Then_Cached() {
+    assertEquals(0, clientCache.estimatedSize());
+
+    final Trait trait1 = environment.client.getTrait(user2, "foo1");
+
+    assertEquals(1, clientCache.estimatedSize());
+    assertNotNull(trait1);
+    assertEquals("foo1", trait1.getKey());
+    assertEquals("xxx", trait1.getValue());
+  }
+
+  @Test(groups = "integration")
+  public void testClient_When_Get_User_Trait_Update_Then_Not_Cached() {
+    final String traitKey = "foo1";
+
+    environment.client.updateTrait(user2, trait(user2.getIdentifier(), traitKey, "new value"));
+    assertEquals(0, clientCache.estimatedSize());
+
+    assertThat(environment.client.getTrait(user2, traitKey))
+        .isEqualTo(trait(null, traitKey, "new value"));
+    assertEquals(1, clientCache.estimatedSize());
+  }
+
+  @Test(groups = "integration")
+  public void testClient_When_Get_User_Trait_Update_Then_Updated_Although_Cached() {
+    // Arrange
+    final String traitKey = "foo1";
+    assignTraitToUserIdentity(user2.getIdentifier(), "unchanged", "stable-value", environment.apiKey);
+
+    assertThat(environment.client.getTrait(user2, traitKey))
+        .isEqualTo(trait(null, traitKey, "xxx"));
```

Comment:
Perhaps I'm missing something but shouldn't the first argument to the `trait` method here be the identifier of `user2` ?

Reply:
(similarly for the rest of the calls to `trait()` which provide `null` as their first argument)

Reply:
Sorry to be a pain but I'm not sure why this one got resolved? I can't see any changes in this area.

Reply:
`assertThat(environment.client.getTrait(user2, traitKey)).isEqualTo(trait(null, traitKey, "xxx"));`

Do you mean why is it checking in the line above that the trait identity is `null`?
The reason is that, that is what the Flagsmith API returns. This is an integration test so it is validating against what the API returns. If you were expecting the identity to be there, then you will need to change the API (it isn't an SDK issue).

You can see that it has been like this for a long time. Tests (that I didn't write nor changed) in FlagsmithClientTest.java have the same behaviour. Check test:

- testClient_When_Add_Traits_For_Identity_Then_Success lines 448, 449, 454, 455
- testClient_When_Add_Traits_For_Identity_To_Existing_Identity_Then_Success lines 468, 469, 474, 475